WIP: Exposure Modes

astronomerdave opened the pull request:
This started as a plan to implement a simple expose function and ended up creating an Exposure Mode base and inherited class structure.

~~Not necessarily to be merged yet, since there's still more to do, but I am creating a PR to solicit input before I continue.~~

Commit:
Moves more controller functions to Controller class and
in doing so addresses comment from PR#83 regarding bias().

astronomerdave committed:
refactor ArchonController::bias() function

Commit:
* provides for instrument-specific sources with instrument-specific
  functions dispatched by the server
* creates a factory function for controllers which eliminates the
  need for camera_interface to know about controller types and
  removes the ifdefs.

astronomerdave committed:
fixes duplicate INSTRUMENT_SOURCES in CMake

astronomerdave committed:
readability improvements

astronomerdave committed:
adds configure_interface() and configure_instrument()

astronomerdave committed:
starts adding functionality needed for expose
fixes bug in ArchonController::send_cmd
adds Archon controller functions to set parameters
branched from DH/instrument-modules, PR#85 not yet approved/merged

astronomerdave committed:
adds get_frame_status and interface to set_parameter

astronomerdave committed:
adds get_parameter

astronomerdave committed:
adds exposure modes

Commit:
set/get exposure modes is working for base and instrument classes

astronomerdave committed:
moves producer/consumer thread responsibility to exposure mode class

astronomerdave committed:
Merge branch 'camera-2.0' into DH/start-simple-exposure

astronomerdave committed:
fixes too many things called expose

prkrtg reviewed (commented):
What are your thoughts on having a central factory for exposure modes?
I thought if we have something like`ExposureModeFactory`, then archon can do something like
`ExposureModeFactory::register_mode<ExposureModeCCD, ArchonInterface>("CCD");`

Same thing in HISPEC, HISPEC developer just does
`ExposureModeFactory::register_mode<ExposureModeFastReadout, Hispec>("FASTREADOUT");`

astronomerdave committed:
replaces "CCD" with "Single" since CCD is not a mode

astronomerdave committed:
adds Camera::Controller::abort and Camera::ArchonController::wait_for_readout

astronomerdave committed:
adds a bunch of ArchonController functions:
print_frame_status, fetch, lock_buffer, unlock_buffer, read_frame
(not compiling at the moment because of read_frame)

astronomerdave committed:
adds to the Camera::Information class and adds functions
load_mode_settings, set_image_geometry to Archon::Controller.
compiles, but untested

Commit:
fixes some bugs with exposure_time and naxes

astronomerdave committed:
adds PARAMs to configure_controller and fixes print_frame_status

Commit:
adds abort, fixes frame status and power control
removes some obsolete files from repo
currently can: open, load, power on, exposuremode, expose, abort

astronomerdave committed:
implements a templated ImageBuffer which the ExposureMode classes inherit

astronomerdave committed:
adds optional args to exposuremode,
calculates geometry and image memory,
producer pushes frames into queue,
adds "test showinfo"

astronomerdave committed:
fixes setting image geometry

astronomerdave commented:
Removed "draft" mode because there is more I want to do but it is getting beyond "expose".
Expose is working.

astronomerdave commented:
> What are your thoughts on having a central factory for exposure modes? I thought if we have something like`ExposureModeFactory`, then archon can do something like `ExposureModeFactory::register_mode<ExposureModeCCD, ArchonInterface>("CCD");`
> 
> Same thing in HISPEC, HISPEC developer just does `ExposureModeFactory::register_mode<ExposureModeFastReadout, Hispec>("FASTREADOUT");`

Where would that go and how would the user make use of it?
Keep in mind, instrument will never change at runtime, but exposure mode certainly can. So changing exposure mode must be something the user can perform at runtime. I realize that some instruments might be fixed in how they operate but other instruments, like NIRC2 for example, will routinely use many different exposure modes with various arguments throughout the night. For example, they will routinely send:

```
CDS 2 10, CDS 2 100, CDS 2, 500, etc.
UTR 100 10, UTR j k, etc.

```

drriddle commented:
My software uses a configuration file to define the exposure modes of a camera (with amplifier settings and other items included in a mode entry).  The software sending the command to set up a camera includes the number for the mode in the command, then the camera software uses the configuration information to set up the camera for that observing mode.  It allows a huge number of camera modes to be used, and doesn't require recompiling the code or complicated coding to make any exposure mode available to the user.

astronomerdave commented:
In this case, an "exposure mode" is compiled code, but it doesn't require **re**compiling to change modes. A defined exposure mode encapsulates what happens when you say "expose". For example, `UTR s k` will contain the steps to delay, read, delay, read, ... for `s` reads, store each read in a slice in a cube, repeat that process `k` times, storing each cube in an extension. Or `MCDS p k` will read `p` times, delay, read `p` times again. Co-add the first group of `p` frames, and subtract from the coadd of the second group. Save that subtracted image. Optionally save all `2*p` frames. Repeat the whole thing `k` times.

There are standard ways to perform UTR, Fowler, etc. and their instructions are encapsulated in a base `exposure_modes` but some instrument may want to do things differently and if so, can override that in their own `<instrument>_exposure_modes`.

An ExposureMode will also contain a DeInterlacing object because different detectors may have different deinterlacing requirements. There are certainly some standard deinterlacing modes, like the DeInterlacing class I've carried around from past projects was readily used for NGPS. NIRC2 is another example of an instrument which had very unique deinterlacing steps.

None of this requires _complicated_ coding, but different exposure modes (or readout modes or sample modes, whatever one calls them) does require  coding.

prkrtg reviewed (approved):
LGTM, great update!
